alf/data_structures.py

```diff
@@ -187,76 +174,89 @@ def restart(observation, action_spec, env_id=None, batched=False):
+        reward (float): A scalar, or 1D NumPy array, or tensor.
+        discount (float): (optional) A scalar, or 1D NumPy array, or tensor.
+        env_id (torch.int32): (optional) A scalar or 1D tensor of 
+            the environment ID(s).
 
     Returns:
         A `TimeStep`.
 
     Raises:
-        ValueError: If observations are tensors but reward's statically known rank
+        ValueError: If observations are tensors but reward's rank
             is not `0` or `1`.
     """
-    first_observation = nest.flatten(observation)
-    first_action = nest.flatten(action)
+    flat_observation = nest.flatten(observation)
     assert all(
         map(torch.is_tensor,
-            first_observation)), ("Elements in observation must be Tensor")
+            flat_observation)), ("Elements in observation must be Tensor")
 
-    # TODO(b/130245199): If reward.shape.rank == 2, and static
-    # batch sizes are available for both first_observation and reward,
+    # TODO: If reward.shape.rank == 2, and static
+    # batch sizes are available for both flat_observation and reward,
     # check that these match.
-    reward = to_tensor(reward, dtype=torch.float32)
+    reward = _to_tensor(reward, dtype=torch.float32)
     assert reward.dim() <= 1, "Expected reward to be a scalar or vector."
     if reward.dim() == 0:
         shape = []
         if env_id is None:
             env_id = torch.tensor(0, dtype=torch.int32)
     else:
-        assert first_observation[0].shape[:1] == reward.shape
-        assert first_action[0].shape[:1] == reward.shape
+        flat_action = nest.flatten(prev_action)
+        assert flat_observation[0].shape[:1] == reward.shape
+        assert flat_action[0].shape[:1] == reward.shape
         shape = reward.shape
         env_id = torch.arange(shape[0], dtype=torch.int32)
     step_type = torch.full(shape, StepType.MID, dtype=torch.int32)
-    discount = torch.tensor(discount, dtype=torch.float32)
+    discount = _to_tensor(discount, dtype=torch.float32)
 
     if discount.dim() == 0:
         discount = torch.full(shape, discount, dtype=torch.float32)
     else:
         assert reward.shape == discount.shape
-    return TimeStep(step_type, reward, discount, observation, action, env_id)
+    return TimeStep(step_type, reward, discount, observation, prev_action,
+                    env_id)
 
 
-def termination(observation, action, reward, env_id=None):
+def termination(observation, prev_action, reward, env_id=None):
     """Returns a `TimeStep` with `step_type` set to `StepType.LAST`.
 
+    Called by env.step() if 'Done'.
+
     Args:
-        observation: A (nested) 'tensor'.
-        reward: A scalar, or 1D NumPy array, or tensor.
+        observation (nested tensors): current observations of the env.
+        prev_action (nested tensors): previous actions to the the env.
+        reward (float): A scalar, or 1D NumPy array, or tensor.
+        discount (float): (optional) A scalar, or 1D NumPy array, or tensor.
```

Comment:
Remove this comment, and maybe add a note that discount shouldn't be used and assumed 0.

Reply:
done